packages/integration/src/github/GithubCredentialsProvider.ts

```diff
@@ -0,0 +1,228 @@
+    }
+
+    if (repositorySelection !== 'all' && !repo) {
+      throw new Error(
+        'Application must be installed for the entire organization',
+      );
+    }
+
+    const cacheKey = !repo ? owner : `${owner}/${repo}`;
+    const repositories = repositorySelection !== 'all' ? [repo!] : undefined;
+
+    // Go and grab an access token for the app scoped to a repository if provided, if not use the organisation installation.
+    return this.cache.getOrCreateToken(cacheKey, async () => {
+      const result = await this.appClient.apps.createInstallationAccessToken({
+        installation_id: installationId,
+        repositories,
+      });
+
+      return {
+        token: result.data.token,
+        expiresAt: DateTime.fromISO(result.data.expires_at),
+      };
+    });
+  }
+
+  private async getInstallationData(owner: string): Promise<InstallationData> {
+    // List all installations using the last used etag.
+    // Return cached InstallationData if error with status 304 is thrown.
+    let installation;
+    try {
+      this.installations = await this.appClient.apps.listInstallations({
+        headers: {
+          'If-None-Match': this.installations?.headers.etag,
+        },
+      });
+
+      installation = this.installations.data.find(
+        inst => inst.account?.login === owner,
+      );
+    } catch (error) {
+      if (error.status !== 304) {
+        throw error;
+      }
+      installation = this.installations?.data.find(
+        inst => inst.account?.login === owner,
+      );
+    }
+    if (installation) {
+      return {
+        installationId: installation.id,
+        suspended: Boolean(installation.suspended_by),
+        repositorySelection: installation.repository_selection,
+      };
+    }
+
+    const notFoundError = new Error(
+      `No app installation found for ${owner} in ${this.baseAuthConfig.appId}`,
+    );
+    notFoundError.name = 'NotFoundError';
+    throw notFoundError;
+  }
+}
+
+// GithubAppCredentialsMux corresponds to a Github installation which internally could hold several GitHub Apps.
+export class GithubAppCredentialsMux {
+  private readonly apps: GithubAppManager[];
+
+  constructor(config: GitHubIntegrationConfig) {
+    this.apps = config.apps?.map(ac => new GithubAppManager(ac)) ?? [];
+  }
+
+  async getAppToken(owner: string, repo?: string): Promise<string | undefined> {
+    if (this.apps.length === 0) {
+      return undefined;
+    }
+
+    const results = await Promise.all(
+      this.apps.map(app =>
+        app.getInstallationCredentials(owner, repo).then(
+          credentials => ({ credentials, error: undefined }),
+          error => ({ credentials: undefined, error }),
+        ),
+      ),
+    );
+
+    const result = results.find(result => result.credentials);
+    if (result) {
+      return result.credentials!.accessToken;
+    }
+
+    const errors = results.map(r => r.error);
+    const notNotFoundError = errors.find(err => err.name !== 'NotFoundError');
```

Comment:
lel

Reply:
I will take full responsibility for this 💩

Reply:
The most joyful part of the entire file ❤️